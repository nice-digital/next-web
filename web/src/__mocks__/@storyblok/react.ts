export const storyblokInit = (): null => null;

export const apiPlugin = (): null => null;
