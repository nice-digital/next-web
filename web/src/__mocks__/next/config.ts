import nextConfig from "./../../../next.config";

const getConfig = (): unknown => ({
	publicRuntimeConfig: nextConfig.publicRuntimeConfig,
	serverRuntimeConfig: nextConfig.serverRuntimeConfig,
});

export default getConfig;
