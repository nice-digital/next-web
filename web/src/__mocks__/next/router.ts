export const useRouter = jest.fn().mockImplementation(() => {
	throw new Error("Provide a mock of useRouter for testing purposes");
});
