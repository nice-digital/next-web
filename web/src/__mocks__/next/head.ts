/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import { FC } from "react";
import { renderToString } from "react-dom/server";

interface HeadProps {
	children: Parameters<typeof renderToString>[0];
}

/**
 * Mocking next/head allows us to use it in our tests to assert against meta tags/titles
 * See https://edibleco.de/3dyJuUl
 */
const Head: FC<HeadProps> = ({ children }: HeadProps) => {
	if (children) {
		global.document.head.insertAdjacentHTML(
			"afterbegin",
			renderToString(children) || ""
		);
	}
	return null;
};

export default Head;
