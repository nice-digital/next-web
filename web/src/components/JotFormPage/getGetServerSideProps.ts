import { GetServerSideProps } from "next";

import { type FormID, getForm } from "@/feeds/jotform/jotform";
import { logger } from "@/logger";

export type FormProps = {
	height: number;
	formID: FormID;
};

// Note: this is a separate file rather than named export from the page component, because the NextJS tree shaking was trying to include needle client side, and causing a build error
export const getGetServerSideProps =
	(formID: FormID): GetServerSideProps<FormProps> =>
	async ({ resolvedUrl }) => {
		try {
			const formResponse = await getForm(
				formID,
				process.env.SERVER_FEEDS_JOTFORM_API_KEY!,
				process.env.PUBLIC_JOTFORM_BASE_URL!
			);

			if (formResponse.responseCode === 404) {
				logger.info(
					`Couldn't find form with id ${formID} at URL ${resolvedUrl}`
				);
				return { notFound: true };
			} else if (formResponse.responseCode === 401)
				throw Error(
					`Got 401 unauthorized response for form with id ${formID} at URL ${resolvedUrl}`
				);

			const { height, status } = formResponse.content;

			if (status === "DISABLED") {
				logger.warn(`Form with id ${formID} at URL ${resolvedUrl} is disabled`);
				return { notFound: true };
			}

			return {
				props: {
					height,
					formID,
				},
			};
		} catch (e) {
			logger.error(
				`Could not get form from JotForm API with id ${formID} at URL ${resolvedUrl}`,
				e
			);
			throw e;
		}
	};
