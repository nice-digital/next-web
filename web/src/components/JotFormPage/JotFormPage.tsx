import { NextSeo } from "next-seo";
import { FC } from "react";

import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { PageHeader } from "@nice-digital/nds-page-header";

import { JotFormEmbed } from "@/components/JotFormEmbed/JotFormEmbed";

import { FormProps } from "./getGetServerSideProps";

export type JotFormPageProps = FormProps & {
	formName: string;
	parentPages: { title: string; path?: string }[];
	lead: string;
	informationPanel?: JSX.Element;
	formHeader?: JSX.Element;
	formFooter?: JSX.Element;
};

export const JotFormPage: FC<JotFormPageProps> = ({
	height,
	formID,
	formName,
	parentPages,
	lead,
	informationPanel,
	formHeader,
	formFooter,
}) => {
	return (
		<>
			<NextSeo
				title={[formName, ...parentPages.map((p) => p.title)].join(" | ")}
				description={lead}
			/>
			<Breadcrumbs>
				{[
					{ title: "Home", path: "/" },
					...parentPages.slice().reverse(),
					{ title: formName },
				].map(({ title, path }) => (
					<Breadcrumb key={title} to={path}>
						{title}
					</Breadcrumb>
				))}
			</Breadcrumbs>
			<Grid gutter="loose">
				<GridItem
					cols={12}
					md={informationPanel ? 8 : 12}
					lg={informationPanel ? 8 : 12}
				>
					<PageHeader heading={formName} lead={lead} id="content-start" />
					{formHeader ? <>{formHeader}</> : null}
					<JotFormEmbed jotFormID={formID} title={formName} height={height} />
					{formFooter ? <>{formFooter}</> : null}
				</GridItem>
				{informationPanel ? (
					<GridItem cols={12} md={4} lg={4}>
						{informationPanel}
					</GridItem>
				) : null}
			</Grid>
		</>
	);
};
