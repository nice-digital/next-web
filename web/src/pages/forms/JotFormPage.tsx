import { NextSeo } from "next-seo";
import { FC } from "react";

import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { PageHeader } from "@nice-digital/nds-page-header";

import { JotFormEmbed } from "@/components/JotFormEmbed/JotFormEmbed";

import { FormProps } from "./JotFormPage.getGetServerSideProps";

export type FormPageProps = FormProps & {
	formName: string;
	parentPages: { title: string; path?: string }[];
	lead: string;
};

export const JotFormPage: FC<FormPageProps> = ({
	height,
	formID,
	formName,
	parentPages,
	lead,
}) => {
	return (
		<>
			<NextSeo
				title={[formName, ...parentPages.map((p) => p.title)].join(" | ")}
				description={lead}
			/>

			<Breadcrumbs>
				{[
					{ title: "Home", path: "/" },
					...parentPages.slice().reverse(),
					{ title: formName },
				].map(({ title, path }) => (
					<Breadcrumb key={title} to={path}>
						{title}
					</Breadcrumb>
				))}
			</Breadcrumbs>

			<PageHeader heading={formName} lead={lead} id="content-start" />

			<JotFormEmbed jotFormID={formID} title={formName} height={height} />
		</>
	);
};
