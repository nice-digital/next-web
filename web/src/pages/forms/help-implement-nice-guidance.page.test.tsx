import { render } from "@testing-library/react";
import { GetServerSidePropsContext } from "next";

import { FormProps } from "@/components/JotFormPage/getGetServerSideProps";

import HelpImplementNiceGuidanceForm, {
	getServerSideProps,
} from "./help-implement-nice-guidance.page";

jest.mock("@/feeds/jotform/jotform", () => ({
	getForm: jest.fn().mockResolvedValue({
		responseCode: 200,
		message: "success",
		content: {
			id: "232203151110028",
			username: "nice_teams",
			title: "Application form",
			height: "539",
			status: "ENABLED",
			created_at: "2022-12-08 06:31:44",
			updated_at: "2023-03-27 05:03:55",
			last_submission: "2023-03-24 14:43:43",
			new: "34",
			count: "46",
			type: "LEGACY",
			favorite: "0",
			archived: "0",
			url: "https://nice.jotform.com/232203151110028",
		},
		duration: "14.98ms",
	}),
}));

describe("HelpImplementNiceGuidanceForm", () => {
	it("should match snapshot", async () => {
		const props = (await getServerSideProps({
			resolvedUrl: "/forms/help-implement-nice-guidance",
		} as GetServerSidePropsContext)) as { props: FormProps };

		const { container } = render(
			<HelpImplementNiceGuidanceForm {...props.props} />
		);

		expect(container).toMatchSnapshot();
	});
});
