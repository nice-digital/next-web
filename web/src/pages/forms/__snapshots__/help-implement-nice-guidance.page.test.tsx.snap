// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpImplementNiceGuidanceForm should match snapshot 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    data-component="breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/implementing-nice-guidance"
        >
          Implementing NICE guidance
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/implementing-nice-guidance/implementation-help-and-advice"
        >
          Implementation help and advice
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/implementing-nice-guidance/implementation-help-and-advice/adoption-and-implementation-support"
        >
          Adoption and implementation support
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Adoption and impact reference panel membership form
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/implementing-nice-guidance",
        "name": "Implementing NICE guidance"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/implementing-nice-guidance/implementation-help-and-advice",
        "name": "Implementation help and advice"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "@id": "/implementing-nice-guidance/implementation-help-and-advice/adoption-and-implementation-support",
        "name": "Adoption and implementation support"
      }
    },
    {
      "@type": "ListItem",
      "position": 5,
      "item": {
        "name": "Adoption and impact reference panel membership form"
      }
    }
  ]
}
  </script>
  <div
    class="grid grid--loose"
  >
    <div
      data-component="grid-item"
      data-g="12 md:12 lg:12"
    >
      <div
        class="page-header "
        data-component="page-header"
        data-testid="content-start"
        id="content-start"
      >
        <h1
          class="page-header__heading "
        >
          Adoption and impact reference panel membership form
        </h1>
        <p
          class="page-header__lead"
        >
          Join the NICE Adoption and Impact Reference Panel. Give your advice on tools and resources as they're being developed. Help us shape the commissioning and delivery of safe, effective and cost-efficient care.
        </p>
      </div>
      <iframe
        allow="geolocation; microphone; camera"
        allowfullscreen=""
        class="iframe"
        data-jotform-id="232203151110028"
        id="JotFormIFrame-232203151110028"
        scrolling="no"
        src="https://next-web-tests.jotform.com/232203151110028?isIframeEmbed=1"
        style="height: 539px; overflow: hidden;"
        title="Adoption and impact reference panel membership form"
      />
    </div>
  </div>
</div>
`;
